### PR TITLE
Issue #2974: EmptyLineSeparator now validates newlines before comments

### LINKED ISSUE #2974: EmptyLineSeparator check does not validate newlines before comments
```
$ javac Test.java
$ cat Test.java
public class Test {

    public static final void method1() { }



    // Method 2 must fail
    public static final void method2() { }



    /**
     * Method 3 must fail
     */
    public static final void method3() { }
}


$ cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
    "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
    "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="EmptyLineSeparator">
            <property name="allowNoEmptyLineBetweenFields" value="true"/>
            <property name="allowMultipleEmptyLines" value="false"/>
        </module>
    </module>
</module>


$ java -jar checkstyle-6.15-all.jar -c config.xml Test.java
Starting audit...
Audit done.
```

**Expected:**
`method2` and `method3` must fail with `... has more than 1 empty lines before. [EmptyLineSeparator]`

Related tickets: #2067 

http://checkstyle.sourceforge.net/config_whitespace.html#EmptyLineSeparator

### PR DESCRIPTION
**Pull request by shivanshsoni**
fixes #2974

### CODE REVIEW

**Comment by rnveach** [comment]
@shivanshsoni Do not create a new PR just to make changes. You can amend the existing branch and commits and push with force option. Github and the CIs will recognize this and re-run everything.
Again, squash all commits and keep them at 1.

**Comment by shivanshsoni** [comment]
@rnveach Sorry Sir, Actually I revert changes of 1st commit i the second commit that's why it has no changes. But I am still working on this issue

**Comment by shivanshsoni** [comment]
@rnveach @romani Why TeamCity build is failing while all other checks are passing? Please suggest some changes to pass all the checks.

**Comment by romani** [comment]
please rebase on our latest code , I just merged a fix for wercker CI to our master.

from Teamcity:
> testAllowMultipleEmptyLinesInsideClassMembers() Contents of array expected are written to, but never read 
 testAllowMultipleEmptyLinesInsideClassMembers() Variable expected is never used

why you have test withtout `verify` method ?

**Comment by rnveach** [comment]
@shivanshsoni Please fix CI so we can continue.

**Comment by liscju** [comment]
@rnveach current version of the code in this PR is nothing but NPE producer:

```java
if (nextToken != null) {
   ...
} else{
   log(nextToken.getLineNo(), MSG_SHOULD_BE_SEPARATED,
         nextToken.getText());
}
```

**Comment by shivanshsoni** [comment]
@rnveach okk sir

**Comment by romani** [comment]
@shivanshsoni , please rebase your code over the most latest our master code to resolve wercker problem.
Please resolve all CI violations.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java [hunk 1]
```diff
@@ -485,9 +485,8 @@ private boolean isPrePreviousLineEmpty(DetailAST token) {
         boolean result = false;
         final int lineNo = token.getLineNo();
         // 3 is the number of the pre-previous line because the numbering starts from zero.
-        final int number = 3;
-        if (lineNo >= number) {
-            final String prePreviousLine = getLines()[lineNo - number];
+        if (lineNo >= 1) {
+            final String prePreviousLine = getLines()[lineNo - 1];
```
Comment (liscju) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: This is incorrect, the comment stated why

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java [hunk 2]
```diff
@@ -545,11 +544,12 @@ private boolean hasEmptyLine(int startLine, int endLine) {
...
         final int lineNo = token.getLineNo();
+        boolean result = false;
         if (lineNo == 1) {
-            return false;
+            return result;
         }
-        //  [lineNo - 2] is the number of the previous line because the numbering starts from zero.
-        final String lineBefore = getLines()[lineNo - 2];
+        //  [lineNo - 1] is the number of the previous line because the numbering starts from zero.
+        final String lineBefore = getLines()[lineNo - 1];
```
Comment (liscju) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: This is wrong, token.getLineNo() returns lines in range <1, N>, getLines operates in lines in range <0, N-1>. So the line returned by token.getLineNo() is in getLines(token.getLineNo() - 1), previous line in getLines(token.getLineNo() - 2). Did you read the comment with understanding?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java [hunk 3]
```diff
@@ -545,11 +544,12 @@ private boolean hasEmptyLine(int startLine, int endLine) {
      */
     private boolean hasEmptyLineBefore(DetailAST token) {
         final int lineNo = token.getLineNo();
+        boolean result = false;
         if (lineNo == 1) {
-            return false;
+            return result;
```
Comment (liscju) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: You extracted the variable result, assigned it to false then returned it so why did you create it in the first place?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java [hunk 4]
```diff
@@ -504,7 +503,7 @@ private boolean hasEmptyLineAfter(DetailAST token) {
             lastToken = token.getLastChild();
         }
         DetailAST nextToken = token.getNextSibling();
-        if (isComment(nextToken)) {
+        if (lastToken != null && isComment(nextToken)) {
```
Comment (liscju) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: You added checking if lastToken is not null but few lines after this lastToken is used to calculate currentEnd:
```java
          if (lastToken != null && isComment(nextToken)) {
                     final int currentEnd = lastToken.getLineNo(); 
                     nextToken = nextToken.getNextSibling();
          }
          // Start of the next token
         final int nextBegin = nextToken.getLineNo();
         // End of current token.
         final int currentEnd = lastToken.getLineNo();
         return hasEmptyLine(currentEnd + 1, nextBegin - 1);
     }
```
This does not make you suspicious that something is clearly wrong?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java [hunk 5]
```diff
@@ -527,7 +526,7 @@ private boolean hasEmptyLine(int startLine, int endLine) {
         boolean result = false;
         if (startLine <= endLine) {
             final FileContents fileContents = getFileContents();
-            for (int line = startLine; line <= endLine; line++) {
+            for (int line = startLine; line < endLine; line++) {
```
Comment (liscju) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java]: Why did you exclude endLine from checking? This is wrong.

**Comment by Vladlis** [comment]
@shivanshsoni , please address @liscju's comments

**Comment by shivanshsoni** [comment]
I think I got it wrong... I will correct all these....

**Comment by liscju** [comment]
@shivanshsoni btw why did you delete branch on your fork with commit in this PR? Are you going to open new PR to fix this issue (and probably piss mainteners off) once again ? :)

**Comment by shivanshsoni** [comment]
@liscju Actually, I rebased new code of checkstyle and after rebasing this branch automatically got deleted. I don't know how...

**Comment by MEZk** [comment]
@shivanshsoni 
Sad, can you restore the changes from the local history of IDEA?

**Comment by romani** [comment]
@shivanshsoni , do you still have a problem ?

**Comment by shivanshsoni** [comment]
yes I am making a new pull request on this issue and closing this one

**Comment by liscju** [comment]
@shivanshsoni i strongly suggest you to do some exercises in git and pull request in your own side repository, because you are doing something clearly wrong. From my investigation this is 4 PR with this single issue that you closed, you were informed multiple times not to close PR and you do it again.

BTW before sending new PR make sure that your changes are fixing something and not breaking existing functionality. So far the changes you sent were not fixing anything, did you even compile them before sending PR?

**Comment by shivanshsoni** [comment]
@liscju Sir, I think I have no need to make any changes in my local repo as whole code got reset when I rebased new code.

**Pull request closed without merging**